File: Assignment.c
/*
Author: Muhammad Danish Zahin
E-mail: mjr7066@psu.edu
Course: CMPSC 460
Assignment: Fortran / C Programming Project
Due date: 2/14/2026
File: Assignment.c
Purpose: This program prompts a user to enter a positive integer, tests if
the integer is perfect, abundant, or deficient, and reports the
result of the test. The program loops until the user decides that
s/he is done testing integers.
IDE/Compiler:
IDE: VScode
Online: TDM64-GCC-10.3.0-2
Download extentsions in VScode:C/C++,modern Fortran, and Code Runner extensions for VScode
Operating system:
MS Windows 11
References: 
How to start coding in Fortran: https://www.youtube.com/watch?v=RrsoM6wVEWE&t=548s 
AI Tool(s): Github Copilot (VScode extension)
Description of code use of pointers and arrays:
In C by default uses pass by value, so we need to use pointers to pass by reference. 
In this program, we will use pointers to allow the functions to modify the values of the variables passed to them 
and change the values of the variables in the main function :).
Array by default is a pointer in C, so we can pass the array to the functions without using pointers.
thank you
*/
#include <stdio.h>

#define MAX_FACTORS 30

// Function prototypes
void getInteger(int *num);
int computeFactors(int num, int factors[]);
void printFactors(int factors[], int count);
int sumFactors(int factors[], int count);
//the main function calling the other functions
int main() {
    int number, factors[MAX_FACTORS], count, sum;
    char choice;
    printf("Welcome to the Perfect, Abundant, and Deficient Number Finder!\n");
    do {    
        printf("----------------------------------------\n");
        getInteger(&number);

        count = computeFactors(number, factors);
        printFactors(factors, count);

        sum = sumFactors(factors, count);

        printf("----------------------------------------\n");
        printf("The sum of the factors is: %d\n", sum);

        if (sum == number) {
            printf("The number %d is perfect.\n", number);
        } else if (sum > number) {
            printf("The number %d is abundant.\n", number);
            printf("Abundance: %d\n", sum - number);
        } else {
            printf("The number %d is deficient.\n", number);
            printf("Deficiency: %d\n", number - sum);
        }
        printf("----------------------------------------\n");

        printf("Do you want to enter another number? (Y/y for yes, any other key for no): ");
        scanf(" %c", &choice);

    } while (choice == 'Y' || choice == 'y');

    return 0;
}
//get integer function to get a positive integer from the user and validate the input
//------------------------------------------------
void getInteger(int *num) {
    int status;
    do {
        printf("Please enter a positive integer: ");
        status = scanf("%d", num);
        while(getchar() != '\n'); // clear input buffer

        if (status != 1 || *num <= 0) {
            printf("Invalid input. Please enter a positive integer.\n");
        } else {
            break;
        }
    } while (1);
}
//compute factors function to compute the factors of the given number and store them in an array, returning the count of factors found
//------------------------------------------------
int computeFactors(int num, int factors[]) {
    int count = 0;
    for (int i = 1; i <= num / 2; i++) {
        if (num % i == 0) {
            factors[count++] = i;
        }
    }
    return count;
}
//print factors function to print the factors of the number stored in the array, handling the case where there are no factors
//------------------------------------------------
void printFactors(int factors[], int count) {
    if (count == 0) {
        printf("The number has no proper factors.\n");
        return;
    }

    printf("The factors of the number are: ");
    for (int i = 0; i < count; i++) {
        if (i < count - 1)
            printf("%d, ", factors[i]);
        else
            printf("%d\n", factors[i]);
    }
}
//sum factors function to calculate the sum of the factors stored in the array and return the sum
//------------------------------------------------
int sumFactors(int factors[], int count) {
    int sum = 0;
    for (int i = 0; i < count; i++) {
        sum += factors[i];
    }
    return sum;
}

File: Assignment.f90

!Author: Muhammad Danish Zahin
!E-mail: mjr7066@psu.edu
!Course: CMPSC 460
!Assignment: Fortran / C Programming Project
!Due date: 2/14/2026
!File: Assignment.f90
!Purpose: This program prompts a user to enter a positive integer, tests if
!the integer is perfect, abundant, or deficient, and reports the
!result of the test. The program loops until the user decides that
!s/he is done testing integers.
!IDE/Compiler:
!VScode
!TDM64-GCC-10.3.0-2
!Download extentsions:C/C++,modern Fortran, and Code Runner extensions for VScode
!Operating system:
!MS Windows 11
!References: 
!How to start coding in Fortran: https://www.youtube.com/watch?v=RrsoM6wVEWE&t=548s 
!AI Tool(s): Github Copilot (VScode extension)
program numberfinder
    IMPLICIT NONE
    INTEGER :: number, factors(30), count, sum
    CHARACTER(LEN=1) :: choice
    print *, "Welcome to the Perfect, Abundant, and Deficient Number Finder!"
    do
        call getInteger(number)
        call computeFactors(number, factors, count)
        call printFactors(factors, count)
        
        sum = sumFactors(factors, count)  

        print *, "The sum of the factors is:", sum
        if (sum == number) then
            print *, "The number is perfect."
        else if (sum > number) then
            print *, "The number is abundant."
            print *, "The abundance is:", sum - number
        else
            print *, "The number is deficient."
            print *, "The deficiency is:", number - sum
        end if

        print *, "Do you want to enter another number? (Enter Y/y for yes, any other key for no)"
        read *, choice
        if (choice /= 'Y' .and. choice /= 'y') then
            exit
        end if
    end do

contains
    ! This subroutine prompts the user to enter a positive integer and validates the input. 
    !It continues to prompt until a valid positive integer is entered. The valid integer is returned through the output parameter 'num'.
    subroutine getInteger(num)
        IMPLICIT NONE
        INTEGER, INTENT(OUT) :: num
        INTEGER :: status
        DO 
            print *, "Please enter a positive integer: "
            read(*, *, IOSTAT=status) num
            if (status /= 0) then
                print *, "Invalid input. Please enter a positive integer."
                cycle
            end if
            if (num > 0) then
                exit
            else
                print *, "Invalid input. Please enter a positive integer."
            end if
        END DO
    end subroutine getInteger
    !This subroutine computes the factors of the given positive integer 'num' and stores them in the output array 'factors'. 
    !The count of factors found is returned through the output parameter 'count'. The subroutine iterates from 1 to num/2 to find all factors of 'num'.
    subroutine computeFactors(num, factors, count)
        IMPLICIT NONE
        INTEGER, INTENT(IN) :: num
        INTEGER, INTENT(OUT) :: factors(30), count
        INTEGER :: i

        count = 0
        do i = 1, num/2
            if (mod(num, i) == 0) then
                count = count + 1
                factors(count) = i
            end if
        end do
    end subroutine computeFactors
    !This subroutine prints the factors of the number stored in the input array 'factors' based on the count of factors provided. 
    !If there are no factors, it prints "None". Otherwise, it iterates through the factors and prints each one.
    subroutine printFactors(factors, count)
        IMPLICIT NONE
        INTEGER, INTENT(IN) :: factors(30), count
        INTEGER :: i

        print *, "The factors of the number are: "
        if (count == 0) then
            print *, "None"
        end if
        do i = 1, count
            print *, factors(i)
        end do
    end subroutine printFactors
    !This function calculates the sum of the factors stored in the input array 'factors' based on the count of factors provided. 
    !It iterates through the factors, adding each one to a running total 'sum', which is returned as the result of the function.
    function sumFactors(factors, count) result(sum)
        IMPLICIT NONE
        INTEGER, INTENT(IN) :: factors(30), count
        INTEGER :: sum, i

        sum = 0
        do i = 1, count
            sum = sum + factors(i)
        end do
    end function sumFactors

end program numberfinder
